File: sdks/go/writer.go
package agentsdk

import (
	"encoding/json"
	"io"
)

// Writer emits JSONL envelopes to an io.Writer with quiet filtering and trace ID injection.
type Writer struct {
	out      io.Writer
	toolName string
	quiet    bool
	traceID  string
}

// NewWriter creates a Writer that emits JSONL lines to out.
func NewWriter(out io.Writer, toolName string) *Writer {
	return &Writer{
		out:      out,
		toolName: toolName,
	}
}

// TraceID returns the current trace ID, or empty string if none is set.
func (w *Writer) TraceID() string {
	return w.traceID
}

// SetQuiet enables or disables quiet mode.
// In quiet mode, progress and warning envelopes are silently dropped;
// result and error envelopes are always emitted.
func (w *Writer) SetQuiet(q bool) {
	w.quiet = q
}

// SetTraceID sets the trace ID injected into all subsequent envelopes.
func (w *Writer) SetTraceID(id string) {
	w.traceID = id
}

// Success emits a result envelope with the given data.
func (w *Writer) Success(data interface{}) error {
	return w.emit(NewResultEnvelope(w.toolName, data))
}

// ErrorWithCode emits an error envelope with the given error code and message.
func (w *Writer) ErrorWithCode(code string, message string) error {
	return w.emit(NewErrorEnvelope(w.toolName, code, message))
}

// Error emits an error envelope with code "error".
func (w *Writer) Error(message string) error {
	return w.emit(NewErrorEnvelope(w.toolName, "error", message))
}

// Warning emits a warning envelope (filtered in quiet mode).
func (w *Writer) Warning(message string) error {
	return w.emit(NewWarningEnvelope(w.toolName, message))
}

// Progress emits a progress envelope (filtered in quiet mode).
func (w *Writer) Progress(percent int, message string) error {
	return w.emit(NewProgressEnvelope(w.toolName, percent, message))
}

// emit is the single bottleneck for all envelope output.
// It applies quiet filtering and trace ID injection.
func (w *Writer) emit(env Envelope) error {
	// Quiet filter: suppress progress and warning in quiet mode.
	if w.quiet && (env.Type == TypeProgress || env.Type == TypeWarning) {
		return nil
	}

	// Inject trace ID if set.
	if w.traceID != "" {
		env.TraceID = w.traceID
	}

	line, err := json.Marshal(env)
	if err != nil {
		return err
	}

	_, err = w.out.Write(append(line, '\n'))
	return err
}

File: sdks/python/agentsdk/writer.py
"""Writer — the single JSONL output bottleneck.

Composes envelope constructors, quiet filtering, and trace ID injection
into a simple API: ``success()``, ``error()``, ``error_with_code()``,
``warning()``, ``progress()``.
"""

from __future__ import annotations

import json
from typing import IO, Optional

from agentsdk.envelope import Envelope


class Writer:
    """Write structured envelopes to a text stream as JSONL.

    Parameters
    ----------
    output:
        Any text-mode ``IO[str]`` (e.g. ``sys.stdout``, ``io.StringIO``).
    tool_name:
        Tool identifier embedded in every envelope.
    """

    def __init__(self, output: IO[str], *, tool_name: str) -> None:
        self._output = output
        self._tool_name = tool_name
        self._quiet: bool = False
        self._trace_id: str = ""

    # ------------------------------------------------------------------
    # Controls
    # ------------------------------------------------------------------

    def set_quiet(self, quiet: bool) -> None:
        """Enable or disable quiet mode.

        When quiet, ``warning()`` and ``progress()`` are silently
        suppressed — no output, no error raised.
        """
        self._quiet = quiet

    def set_trace_id(self, trace_id: str) -> None:
        """Set a trace ID injected into every subsequent envelope."""
        self._trace_id = trace_id

    @property
    def trace_id(self) -> str:
        """Current trace ID (empty string when unset)."""
        return self._trace_id

    # ------------------------------------------------------------------
    # Public emit methods
    # ------------------------------------------------------------------

    def success(self, data) -> None:
        """Emit a **result** envelope carrying *data*."""
        self._emit(Envelope.result(tool=self._tool_name, data=data))

    def error_with_code(self, code: str, message: str) -> None:
        """Emit an **error** envelope carrying *code* and *message*."""
        self._emit(Envelope.error(tool=self._tool_name, error_code=code, message=message))

    def error(self, message: str) -> None:
        """Emit an **error** envelope with the default code ``"error"``."""
        self.error_with_code("error", message)

    def warning(self, message: str) -> Optional[None]:
        """Emit a **warning** envelope. Silently suppressed in quiet mode."""
        if self._quiet:
            return None
        self._emit(Envelope.warning(tool=self._tool_name, message=message))
        return None

    def progress(self, percent: int, message: str) -> Optional[None]:
        """Emit a **progress** envelope. Silently suppressed in quiet mode."""
        if self._quiet:
            return None
        self._emit(Envelope.progress(tool=self._tool_name, percent=percent, message=message))
        return None

    # ------------------------------------------------------------------
    # Internal bottleneck
    # ------------------------------------------------------------------

    def _emit(self, envelope: Envelope) -> None:
        """Inject trace ID (if set), serialize, and write one JSONL line."""
        if self._trace_id:
            envelope.trace_id = self._trace_id
        line = envelope.to_json()
        self._output.write(line + "\n")
